### DISCUSSION TITLE
Python model debugging

### DISCUSSION THREAD

**Discussion by society-research** [FLAMEGPU/FLAMEGPU2#1195] [Q&A] [answered]
Hi there! :)

This question is similar to https://github.com/FLAMEGPU/FLAMEGPU2/discussions/884.

When debuging agent-python..
1. Is it possible to access [printf](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#formatted-output) from agent-python?
2. Any ideas what is wrong with the cuda-gdb workflow described below?

When trying to debug a model written in agent-python using a debug build with `-DFLAMEGPU_RTC_EXPORT_SOURCES=ON` cuda-gdb just hangs forever and nothing happens:
```
(venv) $ cuda-gdb -x gdb.sc --args python3 sx.py -v
(cuda-gdb) break CUDASimulation.cu:simulate()
No symbol table is loaded.  Use the "file" command.
Make breakpoint pending on future shared library load? (y or [n]) y
Breakpoint 1 (CUDASimulation.cu:simulate()) pending.
(cuda-gdb) r
Starting program: /home/ubuntu/model-socix-py/third_party/FLAMEGPU2/build-Debug/venv/bin/python3 sx.py -v
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
starting with: ['sx.py', '-v']
FLAME GPU 2.0.0-rc.2+b0ec5f3e
Simulation configuration:
        Random Seed: 1712743780
        Steps: 1
[New Thread 0x7fff9c1ff640 (LWP 11093)]
[Detaching after fork from child process 11094]
```

I can see some NvDebugAgent running for a while until it stops.
```
/tmp/13414/session1/NvDebugAgent /lib/x86_64-linux-gnu/libcudadebugger.so.1 13414 134 1 0 0 8 9
```

**Comment by Robadob** [comment]
> Is it possible to access [printf](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#formatted-output) from agent-python?

~~I don't believe so.~~

~~You would need to output the codgen'd C agent function source and add them manually.~~

*See @mondus's comment below*

> Any ideas what is wrong with the cuda-gdb workflow described below?

How many agents are in the model you are debugging? CUDA debugging and profiling can be very slow. I see no reason why if your code runs outside the debugger it would fail to run with that particular workflow. @ptheywood is more accomplished with gdb, but he's unavailable till next week.

**Reply by mondus** [reply, marked as answer]
Just to follow up.

> Is it possible to access [printf](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#formatted-output) from agent-python?

Actually yes. #884 Gives an example of how to call `printf` from Python. You will need to use `printf` rather than `print` and treat the arguments as they would be used in C/C++. E.g. provide a string fomratted and the subsequent arguments. This will produce a compilation warning. E.g. "Function call is not a defined FLAME GPU device function or a supported python built in.".

Perhaps a better way is to store the value you want to debug in an agent variable and then either in a step or exit function [obtain the agent population](https://docs.flamegpu.com/guide/running-a-simulation/collecting-data.html#accessing-the-complete-agent-state) so you can examine each of the agent variables.

Not sure why your debugging process is not working but I am not convinced that the run time compilation will produce debug symbols. It certainly wont correspond to your Python source as the code is [transpiled](https://en.wikipedia.org/wiki/Source-to-source_compiler) from Python the C++ before it is compiled.
